Dvclive api reference

daavoo opened the pull request:
> You may disregard these recommendations if you used the **Edit on GitHub** button from dvc.org to improve a doc in place.

❗ Please read the guidelines in the [Contributing to the Documentation](https://dvc.org/doc/user-guide/contributing/docs) list if you make any substantial changes to the documentation or JS engine.

🐛 Please make sure to mention `Fix #issue` (if applicable) in the description of the PR. This causes GitHub to close it automatically when the PR is merged.

Please choose to [allow us](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to edit your branch when creating the PR.

Thank you for the contribution - we'll try to review it as soon as possible. 🙏

Closes https://github.com/iterative/dvclive/issues/108

dberenbaum commented on content/docs/dvclive/api-reference/next_step.md:
```diff
@@ -0,0 +1,43 @@
+# dvclive.next_step()
+
+Signals that the current step has ended and saves _summary metrics_ to a
+`.json`.
```
```suggestion
Signals that the current step has ended and optionally saves _summary metrics_ to a
`.json`.
```

This also can do more, right? Like trigger the html report or make a checkpoint. Specifying everything `next_step` can do in the description might be challenging to maintain. Is there a way to summarize, or maybe refer back to `init` for details? What do you think?

daavoo replied:
Good point. Probably refer back to `init` would be nice

daavoo replied:
Updated with a reference to `init parameters`. Not sure about the current state though as I still specify "everything" in the `Description` section bellow.

daavoo replied:
Would it make sense to move the different behaviors (i.e. enabling/disabling summary/html/checkpoints) to "Example" sections, like in https://dvc.org/doc/api-reference/open#example-use-data-or-models-from-a-dvc-repository ? 

@jorgeorpinel

dberenbaum replied:
That's a good idea. I'm not sure if those belong here or in `init`, but as long as we link between them, either is probably fine.

dberenbaum commented on content/docs/dvclive/api-reference/next_step.md:
```diff
@@ -0,0 +1,43 @@
+
+for step in range(3):
+    dvclive.log("metric", step * step)
+    dvclive.nex_step()
```
```suggestion
    dvclive.next_step()
```

dberenbaum commented on content/docs/dvclive/api-reference/next_step.md:
```diff
@@ -0,0 +1,43 @@
+import dvclive
+
+for step in range(3):
+    dvclive.log("metric", step * step)
```
```suggestion
    dvclive.log("metric", 0.9)
```
For consistency and because it's probably clearer than `step * step`.

dberenbaum commented on content/docs/dvclive/api-reference/log.md:
````diff
@@ -0,0 +1,53 @@
+file.
+
+```py
+ def log(name: str, val: Union[int, float], step: int = None) -> None:
````
Another really minor thought: In the dvc api docs, we omit the return type. Not sure if there's a style preference here, just a thought on consistency. cc @skshetry?

daavoo replied:
I saw the return type being used in get_url and followed that:

https://dvc.org/doc/api-reference/get_url

I should have probably checked the other pages for consistency

dberenbaum reviewed (changes requested):
Great work! This looks really nice as a first draft. I'd also like to get someone from the docs team to review before merging (cc @jorgeorpinel).

We should also open a discussion on whether to include this in docstrings. I know there was already some discussion on autogenerating docs, but I think even if we have to duplicate text, it's worth considering for dvclive. IMO docstrings are generally accepted practice for a lot of ML libraries and could encourage contributions.

daavoo commented on content/docs/dvclive/api-reference/init.md:
```diff
@@ -0,0 +1,61 @@
+
+## Parameters
+
+- `path` (`dvclive` by default) - The _plot metrics_ (generated by
```
I wasn't sure about what term to use. I think I borrowed "plot metrics" from:

https://dvc.org/doc/command-reference/plots#types-of-metrics

I agree that the term could be confusing for new users.

Not sure what "summary plot files" are

daavoo commented:
> We should also open a discussion on whether to include this in docstrings. I know there was already some discussion on autogenerating docs, but I think even if we have to duplicate text, it's worth considering for dvclive. IMO docstrings are generally accepted practice for a lot of ML libraries and could encourage contributions.

I agree that it's worth to add the docstrings to `dvclive` even by copy-pasting content from here. I think I will wait for this P.R. to be merged and after that add the docstrings.

dberenbaum commented:
@jorgeorpinel Should we assign a reviewer from the docs team?

shcheklein commented on content/docs/dvclive/api-reference/init.md:
```diff
@@ -0,0 +1,61 @@
+## Description
+
+It's usage is optional and focused on configuring the behavior of subsequent
+calls to [`dvclive.log`](/doc/dvclive/api-reference/log) and
```
future improvement: we should create a ticket or improve auto-linker, to keep text clean from these internal links like we do in DVC API reference

jorgeorpinel replied:
https://github.com/iterative/dvclive/issues/114

shcheklein commented on content/docs/dvclive/api-reference/init.md:
```diff
@@ -0,0 +1,61 @@
+
+## Parameters
+
+- `path` (`dvclive` by default) - The _metrics logs_ (generated by
```
potential improvement: metrics logs, plots logs - etc - we should link or use "abbr" tooltips to explain them I think

daavoo replied:
I tried to explain those concepts by linking them to `dvc plots` / `dvc metrics` and update some parts for consistency on the naming. Still not 100% convinced, though

shcheklein commented on content/docs/dvclive/api-reference/init.md:
```diff
@@ -0,0 +1,61 @@
+  [`dvclive.next_step`](/doc/dvclive/api-reference/next_step)) will be saved in
+  a single `{path}.json` file.
+
+- `step` (`0` by default) - the `step` values in log files will start
```
do we describe somewhere the format? would be nice to have a link to it

daavoo replied:
Added

shcheklein commented on content/docs/dvclive/api-reference/init.md:
```diff
@@ -0,0 +1,61 @@
+  incrementing from this value.
+
+- `resume` - (`False` by default) - if `True`, DVCLive will try to read the
+  previous step from the path dir and start from that point (unless a `step` is
```
path -> `path`

dir -> directory (or check how do se usually do this in docs)

shcheklein reviewed (changes requested):
some minor comments:

- try to use more links, "abbr" tooltips, for people better understand certain concepts
- write and prioritize the work on the ticket that would allow us to get rid of the explicit MD links to api reference - that should be very easy, similar to DVC.

shcheklein commented:
Btw, on docstrings, I do think we need to include them (way simpler version though).

Also, as a next step, do we want to write some examples for these APIs?

jorgeorpinel commented on content/docs/dvclive/api-reference/log.md:
```diff
@@ -0,0 +1,55 @@
+
+## Description
+
+The first call to `dvclive.log(name, val)` will create a new file in
+`{path}/{name}.tsv` including the header and first row.
```
What is `path` though? Hasn't been explained in this doc. Maybe just link to `init` earlier and avoid this if it's not central to `dvclive.log`

jorgeorpinel commented on content/docs/dvclive/api-reference/log.md:
```diff
@@ -0,0 +1,55 @@
+## Description
+
+The first call to `dvclive.log(name, val)` will create a new file in
+`{path}/{name}.tsv` including the header and first row.
+
+For example `dvclive.log("loss", 0.9)` will create `{path}/loss.tsv`:
```
Also, 💅
```suggestion
`{path}/{name}.tsv` including the header and first row. For example,
`dvclive.log("loss", 0.9)` will create a `dvclive/loss.tsv` file:
```

jorgeorpinel commented on content/docs/dvclive/api-reference/log.md:
````diff
@@ -0,0 +1,55 @@
+
+For example `dvclive.log("loss", 0.9)` will create `{path}/loss.tsv`:
+
+```
````
````suggestion
```tsv
````

jorgeorpinel commented on content/docs/dvclive/api-reference/log.md:
````diff
@@ -0,0 +1,55 @@
+For example `dvclive.log("loss", 0.9)` will create `{path}/loss.tsv`:
+
+```
+timestamp step  loss
+1623671484747 0 0.9
````
Why not preserve the tab characters? It's a TSV file after all.

jorgeorpinel commented on content/docs/dvclive/api-reference/log.md:
````diff
@@ -0,0 +1,55 @@
+```
+
+Each subsequent call to `dvclive.log(name, val)` will add a new row to
+`{path}/{name}.tsv`.
+
+The created file `{path}/{name}.tsv` is usable by `dvc plots`.
````
```suggestion
the TSV file.

> 💡 Note that TSV files can be visualized with `dvc plots`.
```

jorgeorpinel commented on content/docs/dvclive/api-reference/log.md:
```diff
@@ -0,0 +1,55 @@
+
+The created file `{path}/{name}.tsv` is usable by `dvc plots`.
+
+💡 If `name` contains slashes (i.e. `train/loss`), the required subfolders will
+be created and the file will be saved inside the last subfolder (i.e.
+`{path}/train/loss.tsv`).
```
```suggestion
If `name` contains slashes (e.g. `train/loss`), the required directories will
be created (e.g. `dvclive/train/loss.tsv`).
```

jorgeorpinel commented on content/docs/dvclive/api-reference/log.md:
```diff
@@ -0,0 +1,55 @@
+be created and the file will be saved inside the last subfolder (i.e.
+`{path}/train/loss.tsv`).
+
+💡 If you call `dvclive.log` without calling
+[`dvclive.init`](/doc/dvclive/api-reference/init) first, `dvclive` will
+automatically initialize itself using either default values or environment
+variables (when used alongside `DVC`).
```
Again, should probably mention `init` (and this info) at the beginning of the **Description**. And no need to explain DVC environment here as that should be part of the `init()` ref.

jorgeorpinel replied:
Also, we currently use 💡 in a slightly [different manner](https://dvc.org/doc/user-guide/contributing/docs#general-language-guidelines). Up to you but if you want to be consistent with DVC docs for now I'd avoid the emoji here. Could be a note though (md block quote).

jorgeorpinel commented on content/docs/dvclive/api-reference/log.md:
```diff
@@ -0,0 +1,55 @@
+
+## Parameters
+
+- `name` - The _metrics logs_ will be saved in `{path}/{name}.tsv`.
```
- Start with lower case (`- the ...`).
- Not sure about "metrics logs" term
- `path` is out of context.
- No period `.` needed (not a sentence)

jorgeorpinel commented on content/docs/dvclive/api-reference/log.md:
```diff
@@ -0,0 +1,55 @@
+
+- `name` - The _metrics logs_ will be saved in `{path}/{name}.tsv`.
+
+- `val` - The value to be added in the `name` column of a new row.
```
```suggestion
- `val` - the value to add under the `name` column of a new TSV row
```

jorgeorpinel commented on content/docs/dvclive/api-reference/log.md:
```diff
@@ -0,0 +1,55 @@
+
+- `val` - The value to be added in the `name` column of a new row.
+
+- `step` (`None` by default) - The value to be added in the `step` column of a
+  new row. If `None`, the value of the internal `_step` property will be used.
```
I feel like `step` could be explained or at least mentioned in the **Description**? Comes a bit out of the blue here.

jorgeorpinel replied:
Also, I'd avoid term "value" here (can be confused with `val`)

jorgeorpinel commented on content/docs/dvclive/api-reference/next_step.md:
```diff
@@ -0,0 +1,52 @@
+# dvclive.next_step()
+
+Signals that the current step has ended. Check the
+[init parameters](/doc/dvclive/api-reference/init#parameters) for configuring
+the behavior.
```
> Signals that the current step has ended.

What's a step? A step of what?

jorgeorpinel replied:
> Check the init parameters...

This info should probably be in a quick `init()` mention at the beginning of the **Description**, I think (same as I mentioned for `dvclive.log`).

jorgeorpinel commented on content/docs/dvclive/api-reference/next_step.md:
```diff
@@ -0,0 +1,52 @@
+
+## Description
+
+Each call to `dvclive.next_step` will behave depending on the selected
+[init parameters](/doc/dvclive/api-reference/init#parameters) and whether `DVC`
+is available or not.
```
Good Description start here already. But I would just mention briefly that this function uses DVCLive config parameters `summary` and `path` which default to `True` and `dvclive` but can be defined via `init()` (or DVC env but that's also better explained in the `init()` ref.).

Probably no need to complicate this ref with the DVC integration info, as it can be covered in it's own section of the description ⏳

jorgeorpinel commented on content/docs/dvclive/api-reference/init.md:
````diff
@@ -0,0 +1,63 @@
+dvclive.init()
+```
+
+## Description
+
+It's usage is optional and focused on configuring the behavior of subsequent
````
BTW how is `init()` optional? I think an explanation on default values coming from a DVC environment is missing in this ref. And possibly which ones take precedence over the defaults indicated in the **Parameters** list?

jorgeorpinel commented on content/docs/dvclive/api-reference/next_step.md:
````diff
@@ -0,0 +1,52 @@
+}
+```
+
+If `dvclive` is used alongside `DVC`, on each `dvclive.next_step()`, `dvclive`
+will create a [checkpoint](/doc/user-guide/experiment-management/checkpoints).
+In addition, if `html` is True, on each `dvclive.next_step()` call, `DVC` will
+prepare an _html report_ with all the _metrics logs_ logged in `path`.
````
⌛ This is the DVC integration section of the description I was referring to. I think it's best to concentrate all DVC-specific info here. Could even have it's own H3 title if needed.

jorgeorpinel replied:
`_html report_` -> `HTML report`

jorgeorpinel commented on content/docs/dvclive/api-reference/next_step.md:
```diff
@@ -0,0 +1,52 @@
+[init parameters](/doc/dvclive/api-reference/init#parameters) and whether `DVC`
+is available or not.
+
+If `summary` is True, on each `dvclive.next_step()` call, `dvclive` will
+generate a _summary_ of the metrics previously logged with `dvclive.log` and
```
No links?

jorgeorpinel replied:
```suggestion
If `summary` is True, on each `dvclive.next_step()` call, DVCLive will
generate a summary of the values previously logged with `dvclive.log`, and
```

jorgeorpinel commented on content/docs/dvclive/api-reference/next_step.md:
```diff
@@ -0,0 +1,52 @@
+
+If `summary` is True, on each `dvclive.next_step()` call, `dvclive` will
+generate a _summary_ of the metrics previously logged with `dvclive.log` and
+increase the `_step` count.
```
What is `_step`? Looks special with `_`

daavoo replied:
Nothing special, At some point I was thinking to use `_` to indicate that this was an internal property but given that the term `step` is already defined and used everywhere it probably doesn't makes sense

jorgeorpinel commented on content/docs/dvclive/api-reference/next_step.md:
````diff
@@ -0,0 +1,52 @@
+generate a _summary_ of the metrics previously logged with `dvclive.log` and
+increase the `_step` count.
+
+The _metrics summary_ (usable by `dvc metrics`) will be saved to `{path}.json`,
+being `path` the one defined in [dvclive.init](/doc/dvclive/api-reference/init).
+
+The resulting _summary_ of the above code block would be:
+
+```json
+{
+  "step": 2,
+  "metric": 0.9
+}
+```
````
````suggestion
The _metrics summary_ will be saved to `{path}.json`. Here's an example:

```json
{
  "step": 2,
  "metric": 0.9
}
```

>💡 These JSON files can be visualized  `dvc metrics`.

````

jorgeorpinel commented on content/docs/user-guide/experiment-management/checkpoints.md:
```diff
@@ -185,13 +185,14 @@ for i in range(1, EPOCHS+1):
-only code with metadata is versioned in Git (as an ephemeral commit), while the
-actual model weight file will be stored in the DVC data cache.
+You can read about what the line `dvclive.log(k, v)` does in the
+[`dvclive.log`](/doc/dvclive/api-reference/log) reference.
```
```suggestion
[`dvclive.log()`](/doc/dvclive/api-reference/log) reference.
```

jorgeorpinel replied:
Probably other links in all the files are missing `()` too.

daavoo replied:
Copy that. I wasn't sure about whether to use `()` in the links or not

jorgeorpinel reviewed (commented):
Hi 👋 just noticed this. I have lots of comments. Nothing mega major but a few relevant ones, and lots of minor ones (and e.g. copy edits). A couple cosmetic ones 💅 too

Thanks

daavoo commented on content/docs/dvclive/api-reference/index.md:
```diff
@@ -0,0 +1,14 @@
+# API Reference
+
+[DVCLive](https://github.com/iterative/dvclive) is a Python library, which can
+be installed with: `pip install dvclive`.
+
+This reference provides the details about the functions in the API module
+`dvclive`, which can be imported any regular way, for example:
```
Will add it as part of https://github.com/iterative/dvclive/issues/86

daavoo replied:
Sorry I meant https://github.com/iterative/dvclive/issues/116

jorgeorpinel commented:
p.s. @daavoo I realize you may be focused on development so one option is to create a future docs improvement ticket linking to my review above. Maybe we can even work on it ourselves (docs team) — we can discuss in a planning meeting. Tanks

jorgeorpinel commented:
I guess could be part of iterative/dvclive/issues/86 but that's already a pretty long one 😅 kind of an epic?

daavoo commented:
> p.s. @daavoo I realize you may be focused on development so one option is to create a future docs improvement ticket linking to my [#2632 (review)](https://github.com/iterative/dvc.org/pull/2632#pullrequestreview-708848857) above. Maybe we can even work on it ourselves (docs team) — we can discuss in a planning meeting. Tanks

I was already addressing your suggestions locally. My personal opinion is that, for `dvclive`, extending existing documentation would be more valuable at this point than development of new features and that is also what we have agreed at our planning.

daavoo commented:
> UPDATE: I guess could be part of [iterative/dvclive/issues/86](https://github.com/iterative/dvclive/issues/86) but that's already a pretty long one sweat_smile kind of an epic?

Yep. I tried to use the Github `tasks` feature to create new issues from the task list. Will create a new one for this `API Reference` updates.

daavoo commented on content/docs/dvclive/api-reference/init.md:
```diff
@@ -0,0 +1,63 @@
+  more details.
+
+- `resume` - (`False` by default) - if `True`, DVCLive will try to read the
+  previous `step` from the `path` directory and start from that point (unless a
+  `step` is passed explicitly).
```
`step = 0` is the default value so passing `resume=True` will behave just like  `step=0, resume=True`. 

The behavior will be that `dvclive` will try to read the latest `step` stored in the *summary metrics* file. If the value stored in *summary metrics* is other than `0`, `dvclive` will ignore the passed `step=0`:

https://github.com/iterative/dvclive/blob/master/dvclive/metrics.py#L37

daavoo replied:
I added a clarification for that

shcheklein commented:
@jorgeorpinel to clarify, @daavoo 's scope of work includes writing docs/README and other product things, not only engineering. It's expected that he takes of the product holistically (similar to Casper's work on CML I think).

jorgeorpinel commented:
Got it, thanks guys.
